diem-node/src/lib.rs

```diff
@@ -497,12 +543,13 @@ pub fn setup_environment(node_config: &NodeConfig, logger: Option<Arc<Logger>>)
         .spawn(periodic_state_dump(node_config.to_owned(), db_rw));
 
     DiemHandle {
-        _network_runtimes: network_runtimes,
-        _mempool: mempool,
-        _state_sync_bootstrapper: state_sync_bootstrapper,
+        _api: api_runtime,
+        _backup: backup_service,
         _consensus_runtime: consensus_runtime,
         _debug: debug_if,
-        _backup: backup_service,
-        _api: api_runtime,
+        _mempool: mempool,
+        _network_runtimes: network_runtimes,
+        _state_sync_bootstrapper: state_sync_bootstrapper,
+        _storage_service: storage_service_rt,
```

Comment:
Man one of these days we need to fix the explosion of runtimes we have here

Reply:
yeah, I think in an ideal world we would need only one tokio runtime. I'm only speculating here, but having N runtimes each running `num_cores` threads probably leads to a lot of unnecessary thread context switching, though we should profile to be sure : p

unfortunately, the reality of our codebase is that we aren't super disciplined about avoiding blocking/expensive code in Futures/tasks across our various applications, so the "many runtimes" strategy at least provides some basic quality-of-service/fairness guarantees across applications.

i think if we really wanted to be disciplined, we would use tokio-tracing, annotate every spawned task with a human readable name, and then record the "time spent polling" vs "total task lifetime" in our shared metrics. any tasks blocking for longer than ~5 ms should probably get spawned on the blocking CPU thread pool.

Reply:
I remember playing around with this a while back and the end result was that a single shared runtime across everything resulted in less consensus throughput because consensus wasn't being given as much CPU time. It's possible we might be able to achieve an ideal trade-off though, where we have a limited number of runtimes (e.g., to ensure higher priority components, e.g., consensus, are given more CPU time. But, right now, it looks like every component spawns its own runtime 😄